### PR TITLE
Saving themes implemented and packaged the project into single module

### LINKED ISSUE #18: Saving Themes
**Is your feature request related to a problem? Please describe.** 
When opening the application, it always opens in dark mode, even though I may have set the application to light mode previously.  

**Describe the solution you'd like** 
When I open the application, I would like the theme to always be the one I used last. If I open the application, set the theme to light mode, close the program and reopen it, I would expect it to stay in light mode. Likewise, if I keep it in dark mode, I would always expect the application to be in dark mode. 

I would like the theme to be saved into a JSON file in the repo. 

**Describe alternatives you've considered** 
The theme could be saved into a CSV file, but I think JSON would be easier to handle (based on experience) and would be easier to scale and maintain in the future. Feel free to disagree with me here though! 

Another alternative could be a database, but I think that would be a bit over-the-top at this stage.  

This issue is open to being worked on for Hacktoberfest, comment below if you have any questions and/or would like to work on it! First come first serve.

### PR DESCRIPTION
**Pull request by siddharth595**
**Issue Number: #18**

**Added the feature of saving the theme of the previously opened session using a JSON file. The JSON file also stores different theme configurations which can be extended as and when required. Thus the application reads the theme from the JSON file instead of the hard-coded format inside the application. Also combined all the python files into a single folder(module) and it can be run using the run.py file(like an abstraction). Also deleted a redundant setup.py file.**

image
image

** @cBridges851  Please let me know if any changes are required as I am a beginner to open source**

### CODE REVIEW

File: Image_Viewer/file_handler.py
Comment (cBridges851) [inline Image_Viewer/file_handler.py:2]: Can you remove the `__pycache__` folder please? This is my bad - I didn't originally have a .gitignore file in the repo (if you've not come across a .gitignore file, it's just a file that lists the files that shouldn't be picked up by Git when any changes are made to it or files have been added to a specified directory). Files that are in the __pycache__ folder should be ignored as every time you run the application, they will change, which will fog up what needs to be reviewed in PRs and they don't impact the actual running of the application.

A .gitignore file was added to the repo a few days ago. If you remove the `__pycache__` folder, pull latest master, and run the application, these files won't show up in Git again :)

**Comment by cBridges851** [comment]
Can you remove the __pycache__ folder please? This is my bad - I didn't originally have a .gitignore file in the repo (if you've not come across a .gitignore file, it's just a file that lists the files that shouldn't be picked up by Git when any changes are made to it or files have been added to a specified directory). Files that are in the pycache folder should be ignored as every time you run the application, they will change, which will fog up what needs to be reviewed in PRs and they don't impact the actual running of the application.

A .gitignore file was added to the repo a few days ago. If you remove the __pycache__ folder, pull latest master, and run the application, these files won't show up in Git again :)

File: Image_Viewer/menu.py
Comment (cBridges851) [inline Image_Viewer/menu.py:23]: Here's an idea around the arguments of `change_theme(...)` which isn't what I'd consider to be a blocker to merging, but I'll share it with you anyway:

`"Dark"` and `"Light"` could be considered as "magic strings", meaning it's a string that is defined specifically in the application code. These strings have to match the key in the JSON file.

We may need to do `change_theme("Dark")` elsewhere (for example, there could be another route the user could take to change their theme in the future). If we changed the key in the JSON file, we would have to do a find and replace to change any occurrences of that "Dark" string. There could also be problems if a typo is made in the string that is passed into `change_theme()`, as it would mean that the key isn't found.

What we could do is create a class that contains constant strings for the themes, it could look something like this (the caps are convention):
image

This means that if a change needed to be made to a string, that can be done in one place, rather than trying to find all occurrences of that string.

To use that constant string, you'd do this:
image
(Note: I left the label deliberately in that example as that's a UI string, which is being accounted for in #28, so you wouldn't need to worry about that, and I wouldn't put UI strings in a constant strings file as that would make things a bit challenging when it comes to adding new languages).

I'll leave it up to you as to whether you make this change (or what other ways magic strings can be handled!), I personally think it's nice to have and would set a good example to any future collaborators!

File: Image_Viewer/renderer.py
Comment (cBridges851) [inline Image_Viewer/renderer.py:1]:
```suggestion

```
This can be removed since we're not using it

File: Image_Viewer/theme.py
Comment (cBridges851) [inline Image_Viewer/theme.py:2]:
```suggestion

```
This isn't used so can be removed

File: Image_Viewer/theme.py
Comment (cBridges851) [inline Image_Viewer/theme.py:8]: Tiny comment: There's a lot of space between the `__init__` and `set_theme` methods, that can be removed - just need a one line space

File: Image_Viewer/theme.py
Comment (cBridges851) [inline Image_Viewer/theme.py:26]: I'd suggest putting these onto separate lines just to help make it a bit easier to read:
```suggestion
        themeDictionary = {
            "themes": self.themes,
            "lastTheme": self.current_theme
         }
```

Comment (cBridges851) [inline Image_Viewer/theme.py:26]: I think we can improve this writing bit further - the only thing that changes is what the user's last theme was.

I think it would be a good idea to:
1) Keep all the themes in themes.json, I really like having the themes out of the code. It'll make it super easy to add a new theme.
2) Have "lastTheme" in a separate JSON file, maybe something like customsettings.json to open up possibilities for further saved settings in the future. That customsettings.json file could be added to the gitignore (which is currently just on latest master), so it wouldn't be picked up in Git.

Arguably, if there are more settings in the future, that could mean we run into the same thing around changing a whole file when only one thing is updated, but I think the difference here is that the themes themselves are constant, whereas user preferences aren't, so I think it's better to separate them out.

How does this sound?

File: Image_Viewer/menu.py
Comment (cBridges851) [reply Image_Viewer/menu.py:23]: Good thinking! I'll let you implement whichever you think would be best :)

**Comment by siddharth595** [comment]
@cBridges851  I have made the suggested changes. Let me know if further improvements are required

image
image

image

**Comment by cBridges851** [comment]
> @cBridges851 I have made the suggested changes. Let me know if further improvements are required
> 
> image image
> 
> image

That was really quick! Thank you!

**Comment by cBridges851** [comment]
Since you now have to run `run.py` to launch the application, and the screenshot of the UI has been moved, can you update the README please?
image

I think I'll put a note on the PR template around updating the README if necessary for any future contributors, it's something I forget about myself 😂. Don't worry about updating anything else that's out of date as that's beyond this PR, I'll deal with that 😊.

**Comment by cBridges851** [comment]
Heya @siddharth595! Another PR has just been merged which has caused some merge conflicts on this PR. If you pull (or maybe fetch) latest master from cBridges851/master, your IDE should show where the conflicts are and you can select to keep the incoming changes, your changes, or both. Which one you'll need to accept will depend on what it is. The changes you should see are around strings that are displayed in the UI being moved out to a JSON, as well as some other strings around the display of the application (excluding themes).

Feel free to let me know if you need any more guidance around that, I'm more than happy to help :)

There's also two more comments above around a remaining unresolved comment and the README - this PR isn't far off merging!

Thank you :)

**Comment by siddharth595** [comment]
> Since you now have to run `run.py` to launch the application, and the screenshot of the UI has been moved, can you update the README please? image
> 
> I think I'll put a note on the PR template around updating the README if necessary for any future contributors, it's something I forget about myself 😂. Don't worry about updating anything else that's out of date as that's beyond this PR, I'll deal with that 😊.

HI @cBridges851, working on this.

File: Image_Viewer/theme.py
Comment (siddharth595) [reply Image_Viewer/theme.py]: Sorry, missed this part due to a series of comment, will implement this.

**Comment by siddharth595** [comment]
Hi @cBridges851, I have resolved the merge conflicts. I have also made the further changes as asked in the above comments.
1) Updated readme:
image

2) Also separated "lastTheme" from the theme list
image
image

3)
image
Also the following functions have been modified. The *read_custom_settings()* function has kept in Theme class for now as there are no other current utitlities of customSettings.json. In future, when more configurations are added in the json another custom class will be required.

@cBridges851 please review this at the earliest.

**Comment by siddharth595** [comment]
Also please check if the application supports jfif images. I tried some, but they were not being displayed.

**Comment by cBridges851** [comment]
> Hi @cBridges851, I have resolved the merge conflicts. I have also made the further changes as asked in the above comments.
> 
> 1. Updated readme:
>    image
> 2. Also separated "lastTheme" from the theme list
>    image
>    image
> 
> image Also the following functions have been modified. The _read_custom_settings()_ function has kept in Theme class for now as there are no other current utitlities of customSettings.json. In future, when more configurations are added in the json another custom class will be required.
> 
> @cBridges851 please review this at the earliest.😊

Thank you @siddharth595, I really appreciate how clear you make the changes in these comments! I'm taking a look now :)

**Comment by cBridges851** [comment]
Minor comment: I'd suggest for the `customSettings.json` and `themes.json` to be put in the settings folder

**Comment by cBridges851** [comment]
> Also please check if the application supports jfif images. I tried some, but they were not being displayed.

jfif isn't currently supported in the application, this screenshot below is from file_handler.py, which is where the supported file types are specified, which is why jfifs aren't showing up. 
image

They should be easy to add in, but don't worry about doing that in this PR. I'll create another issue for that which you're welcome to pick up if you like :)

**Comment by siddharth595** [comment]
> Minor comment: I'd suggest for the `customSettings.json` and `themes.json` to be put in the settings folder

I have made the required change @cBridges851, please let me know if more changes are required.

**Review by cBridges851 — APPROVED** [review]
Thank you for the changes you've made here @siddharth595, you've done a fantastic job, I'm one very happy maintainer! 

You've made this PR really easy to review via the screenshots you put in the PR descriptions and responses to comments, and you handled the comments themselves really well - do keep this up on your open source journey! 

As for the solution itself, the way you've implemented it is very easy to understand and you've helped improve the layout of the codebase!

This PR is a perfect example of great communication and initiative - it's been a huge pleasure working with you on this!


Given all of what I've said above, I'd love to share what you've done here on my LinkedIn, would you be happy for me to do that? You'll see on my activity the posts I've done for previous contributors to get an idea of what I'm writing (https://www.linkedin.com/in/christa-bridges/). I can also share your GitHub profile if you like.

Thank you and all the best with your open source journey, keep up what you're doing, other maintainers will really appreciate it! You're also very welcome to contribute here again, feel free to let me know if there's anything you're looking to upskill on, and I'm more than happy to see if I can find an opportunity for that within this repo or another repo I have.

**Comment by siddharth595** [comment]
Thankyou @cBridges851 for your kind words and giving me an opportunity to contribute. You are a great project maintainer and know how to take a project forward and improve it step by step. I would like to contribute to your projects often and learn a lot from your projects. Feel free to share our work on LinkedIn. 😊

**Comment by cBridges851** [comment]
Thank you @siddharth595, I really appreciate that, I'll keep striving to be the best project maintainer I can be! It would be great to see you contribute to my projects again! 
I've just shared the LinkedIn post! https://www.linkedin.com/posts/christa-bridges_softwareengineering-opensource-hacktoberfest-activity-6992448192139116544-UJxO?utm_source=share&utm_medium=member_desktop

Thanks so much again 😁

**Pull request merged**
